src/Data/Singletons/Single/Monad.hs

```diff
@@ -154,3 +171,66 @@ singDecsM :: DsMonad q => [Dec] -> SgM [DDec] -> q [DDec]
 singDecsM locals thing = do
   (decs1, decs2) <- singM locals thing
   return $ decs1 ++ decs2
+
+{-
+Note [Explicitly binding kind variables]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+singletons is careful to generate code that explicitly quantifies every kind
+variable bound by a top-level forall. For example, if we were to single the
+identity function:
+
+  identity :: forall a. a -> a
+  identity x = x
+
+We would first promote the types to kinds and generate an explicit forall
+mapping fresh type variables to the promoted types (step 1):
+
+  sIdentity :: forall (x :: a). Sing x -> Sing (Identity x)
+  sIdentity sX = sX
+
+Now, because we want to explicitly bind the kind variable `a`, we traverse
+the promoted types, extract the set of all kind variables, and put
+them at the front of the forall, like so (step 2):
```

Comment:
Actually, the first approach might be considerably simpler. You can do it all in the promotion pass, making the implicit quantification explicit in the `ADxxx` AST. Then, when singling, you just have to preserve explicit quantification. Still, I'm not asking you to rewrite this.